### ISSUE TITLE
official chef yum repo looks broken (2.8 gb on disk)

### ISSUE THREAD

**Issue by prudnitskiy** [chef/chef#7331]
## Description
Official chef repository for centos 7 takes too much disk, it is not looks normal:

```
[root@db1 ~]# ls -alh /var/cache/yum/x86_64/7/chef/gen/
total 2.8G
drwxr-xr-x. 2 root root 4.0K May 31 23:01 .
drwxr-xr-x. 4 root root 4.0K Jun  4 11:05 ..
-rw-r--r--. 1 root root 2.5G May 30 10:17 filelists.xml
-rw-r--r--. 1 root root 134M May 31 16:12 filelists.xml.sqlite
-rw-r--r--. 1 root root  87K May 30 10:17 other.xml
-rw-r--r--. 1 root root  61K May 31 16:12 other.xml.sqlite
-rw-r--r--. 1 root root  65M May 31 21:50 primary.xml
-rw-r--r--. 1 root root 135M May 31 23:01 primary.xml.sqlite
```
 
## Chef Version
chef 14

## Platform Version
centos 7 x86_64

## Replication Case
add yum repo and run `yum makecache`

PS I can't find an appropriate repo to place this ticket, if I missed - please direct me to appropriate repo, and I will reopen the ticket.

**Comment by lamont-granquist** [comment]
yeah its not really the right issue, but IDK what the right one is either, i passed it along via internal slack.

**Comment by sclark007** [comment]
We have been dealing with this issue for the chef-yum repo for a couple weeks too. looks like the chef-stable filelists are 2.6G and this was filing up our /var mounts. 

[root@default-centos-7 7]# du -ksh /var/cache/yum/x86_64/7/chef-stable/gen/*
2.6G	/var/cache/yum/x86_64/7/chef-stable/gen/filelists.xml
1.2G	/var/cache/yum/x86_64/7/chef-stable/gen/filelists.xml.sqlite

when comparing to epel yum repo
[root@default-centos-7 7]# du -ksh /var/cache/yum/x86_64/7/epel/gen/*
140M	/var/cache/yum/x86_64/7/epel/gen/filelists.xml
64M	/var/cache/yum/x86_64/7/epel/gen/filelists.xml.sqlite

**Comment by jaymalasinha** [comment]
@prudnitskiy @sclark007 The unzipped filelists.xml is 2.6G.  You can gzip the cached file to reduce it to 145M 

```
[root@default-centos-7 ~]# du -sh /var/cache/yum/x86_64/7/chef-stable/gen/*
145M	/var/cache/yum/x86_64/7/chef-stable/gen/filelists.xml.gz
1.2G	/var/cache/yum/x86_64/7/chef-stable/gen/filelists.xml.sqlite
92K	/var/cache/yum/x86_64/7/chef-stable/gen/other.xml
68K	/var/cache/yum/x86_64/7/chef-stable/gen/other.xml.sqlite
67M	/var/cache/yum/x86_64/7/chef-stable/gen/primary.xml
138M	/var/cache/yum/x86_64/7/chef-stable/gen/primary.xml.sqlite

[root@default-centos-7 ~]# yum -C list chef-14*
Loaded plugins: fastestmirror
Available Packages
chef.x86_64                      14.3.37-1.el7                       chef-stable
```

**Comment by irvingpop** [comment]
So, a few notes on this:
- Chef omnibus packages are virtually self-contained Linux distributions (minus the kernel and glibc), so they tend to be quite large; especially our server packages.  This is by design and provides many benefits in terms of software quality and testability.  So it's not entirely fair to compare our packages to those of epel or even Redhat, which focus on packaging the smallest possible unit.
- Yum does some less-than-ideal things architecturally,  like storing both the compressed and decompressed copies of the metadata.  I did a read through the code and this doesn't even seem configurable.  Perhaps there are yum plugins that can solve for this.
- You don't need to `yum makecache` to retrieve packages from the repository.   In fact, the filelist isn't necessary for basic yum operations like `yum install` and `yum upgrade`.  In my experiments from a clean cache, only `primary.xml.*` are downloaded which only takes up 208 MB in aggregate
- We're storing deprecated packages in the `-stable` channel that we should probably move off to a different channel, like `-archive` or something.   For example, take a look at `yum list --showduplicates chefdk`  (or any of `automate|delivery|supermarket|etc...`)

## general advice to customers:
- If you're somewhat space constrained, consider skipping the `yum makecache` phase.   package installs/upgrades will still work, and you'll only store 1/20th the amount of metadata on disk.
- For extreme space savings, consider building your own yum repos consisting of only the client packages/versions you use.  It wouldn't be hard to build a script that queries available versions from omnitruck, fetches new versions and then builds a yum repo that can be served off of Apache or something.

## Things we (Chef) could do better:
- Purge deprecated packages from the stable channel out to some kind of archive-type channel.  That could save us on CDN costs by serving smaller metadata to customers, and serving the archive channel from slower/lower-availability systems. 

## in conclusion
Our yum repos are not broken.  They're just large.  I recommend closing this issue.

**Comment by coderanger** [comment]
@irvingpop All the listed things weren't about package sizes, that's metadata files. It seemed like filelists.xml probably shouldn't be quite that big, even with our very big packages :)

**Comment by coderanger** [comment]
Fair, it's reasonable to say we just won't fix it, but I think we should consider maybe zapping those packages from the repo if it's just a few packages that got mis-built or something :)

**Comment by jaymalasinha** [comment]
Thank You @irvingpop for the excellent detailed report, recommendations and conclusion on this issue.  Closing it now.

**Comment by sclark007** [comment]
@irvingpop when you compare the rpm counts to say epel or even centos that have thousands of rpm's in them there is still a issue with the yum-chef being that size. for CIS type builds we have always used a /var filesystem of 4G for year with no issue until recently.

**Comment by bish0polis** [comment]
We found a similar-sounding issue very early on, but we had a bad interaction ("it's on you") and never raised this one.  We worked around this issue with a cobbler repo-cache install:  We were able to direct cobbler to cache only the most recent version of any package, removing dupes, and it kept the downstream yum caches small by regenerating the meta-data only on a single stable RPM of each.

It may seem a bit extreme to light up a cobbler install for one repo, but we had one inside our spacewalk server already; and if you have one set up at your site it could be a quick config change.
